packages/builder/src/package.ts

```diff
@@ -112,6 +112,48 @@ export async function copyPackage({ packageRef, tags, variant, fromStorage, toSt
   }
 }
 
+/**
+ * Recursively reads the given package from IPFS and makes sure to pin it on the publishStorage.
+ */
+export async function publishIpfs({
+  packageRef,
+  chainId,
+  preset,
+  readStorage,
+  publishStorage,
+}: {
+  packageRef: string;
+  chainId: number;
+  preset: string;
+  readStorage: CannonStorage;
+  publishStorage: CannonStorage;
+}) {
+  const deployData = await readStorage.readDeploy(packageRef, preset, chainId);
+
+  if (!deployData) {
+    throw new Error('ipfs could not find deployment artifact. please double check your settings, and rebuild your package.');
+  }
+
+  const metaUrl = await readStorage.registry.getMetaUrl(packageRef, `${chainId}-${preset}`);
+
+  const copyIpfs = async (deployInfo: DeploymentInfo) => {
+    const newMiscUrl = await readStorage.putBlob(await readStorage.readBlob(deployInfo.miscUrl));
+
+    let newMetaUrl = metaUrl;
+
+    if (metaUrl) {
+      newMetaUrl = await publishStorage.putBlob(await readStorage.readBlob(metaUrl));
+      if (!newMetaUrl) throw new Error('error while writing new misc blob');
+    }
```

Comment:
Should we save the new MetaUrl somewhere in the copy deployInfo, to upload again?